### PR TITLE
Expose an ESM exclusive binary

### PR DESCRIPTION
**Pull request by dvnrsn**
- create bin.[cjs, mjs]
- duplicate tests for mjs
- duplicate *.cjs files used in binode testing examples

### CODE REVIEW

File: bin.mjs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+} else {
+  binName = pkgName = binPkg;
+}
+
+const modPkgPath = resolve.sync(`${pkgName}/package.json`, {
+  basedir: process.cwd(),
+});
+const modPkgDir = path.dirname(modPkgPath);
+
+fs.readFile(modPkgPath, 'utf8')
```
Comment (dvnrsn) [inline bin.mjs:27]: This is the main change. Not as easy as a "require" to read JSON in ESM.

File: test/index.mjs [hunk 2]
```diff
@@ -0,0 +1,119 @@
+import { spawnSync } from 'child_process';
+import path from 'path';
+import { fileURLToPath } from 'url';
```
Comment (dvnrsn) [inline test/index.mjs:3]: This test is virtually the same except it runs `run-binode*-esm` in the package.json which in turn uses the `binode-esm` binary

File: test/check-global.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+if (!global.globalSetup) {
```
Comment (dvnrsn) [inline test/check-global.js:1]: Maybe overkill to duplicate these files but wanted to prove the full ESM format case

File: package.json [hunk 4]
```diff
@@ -2,19 +2,19 @@
...
-    "test": "node test"
+    "test": "node test/index.mjs && node test/index.js"
   },
   "keywords": [],
   "author": "Kent C. Dodds <me@kentcdodds.com> (https://kentcdodds.com)",
   "license": "MIT",
   "bin": {
-    "binode": "bin.js"
+    "binode": "bin.cjs",
+    "binode-esm": "bin.mjs"
```
Comment (dvnrsn) [inline package.json:13]: Could be clever to simply tease out which module system is being used and proceed from there. Seems like that could be doable I suppose by checking `import.meta` or __dirname. Haven't investigated

**Comment by kentcdodds** [comment]
Thanks for this. Can you help me understand why this is necessary?

**Comment by dvnrsn** [comment]
I wanted to use remix-utils. But remix-utils@7 prefers `type: module` [if I understand correctly](https://github.com/sergiodxa/remix-utils#usage-with-cjs).

I authored this proposal under the impression that an ESM exclusive binary was necessary in a `type: module` context. I came to that impression because of my own inexperience migrating remix indie-stack from CJS to MJS and the comment thread linked in this issue https://github.com/kentcdodds/binode/issues/5.

After noticing that for some reason `NODE_OPTIONS=\"--require ./mocks\" remix-serve ./build/index.js` is [breaking my debugger](https://discord.com/channels/770287896669978684/1174056183787704431/1174056183787704431), I thought my only alternative was binode in ESM.

However, as I noted in the [same issue](https://github.com/kentcdodds/binode/issues/5) mentioned earlier, I believe using `--import` is permissible, even though it is experimental. In fact, I'm not sure how to otherwise run binode in an ESM context without using `--import`. This seems to work for me.

At this point I simply want to do whatever I can to [get indie-stack to a better state](https://github.com/remix-run/indie-stack/pull/255#issuecomment-1809400072) for newcomers (like me). It might be best to pocket mock service worker as an option for down the road, for example.

Or maybe people should be [pointed to templates](https://github.com/sergiodxa/remix-utils/discussions/274#discussioncomment-7537148)?

**Pull request closed without merging**